File: Listas, Colas y Pilas/Pilas/PilaListas.java
class Nodo {
    int data;
    Nodo next;

    public Nodo(int d) {
        data = d;
        next = null;
    }
}

class PilaLista {
    private Nodo top;

    public boolean isEmpty() {
        return top == null;
    }

    public void push(int data) {
        Nodo nuevo = new Nodo(data);
        nuevo.next = top;
        top = nuevo;
    }

    public int pop() {
        if (isEmpty()) {
            System.out.println("Error: Stack Underflow");
            return -1;
        }
        int dato = top.data;
        top = top.next;
        return dato;
    }

    public int peek() {
        if (isEmpty()) {
            System.out.println("Pila vacía");
            return -1;
        }
        return top.data;
    }

    public static void main(String[] args) {
        PilaLista pila = new PilaLista();
        pila.push(10);
        pila.push(20);
        pila.push(30);

        System.out.println("Elemento superior: " + pila.peek());
        System.out.println("Extrae elemento: " + pila.pop());
        System.out.println("Nuevo elemento superior: " + pila.peek());
    }
}

File: Listas, Colas y Pilas/Pilas/PilaListas.cpp
#include <iostream>
using namespace std;

class Nodo {
public:
    int data;
    Nodo* next;

    Nodo(int d) {
        data = d;
        next = nullptr;
    }
};

class PilaLista {
private:
    Nodo* top;

public:
    PilaLista() {
        top = nullptr;
    }

    bool isEmpty() {
        return top == nullptr;
    }

    void push(int data) {
        Nodo* nuevo = new Nodo(data);
        nuevo->next = top;
        top = nuevo;
    }

    int pop() {
        if (isEmpty()) {
            cout << "Error: Stack Underflow\n";
            return -1;
        }
        int dato = top->data;
        Nodo* temp = top;
        top = top->next;
        delete temp;
        return dato;
    }

    int peek() {
        if (isEmpty()) {
            cout << "Pila vacía\n";
            return -1;
        }
        return top->data;
    }
};

int main() {
    PilaLista pila;
    pila.push(10);
    pila.push(20);
    pila.push(30);

    cout << "Elemento superior: " << pila.peek() << endl;
    cout << "Extrae elemento: " << pila.pop() << endl;
    cout << "Nuevo elemento superior: " << pila.peek() << endl;

    return 0;
}
